### PR TITLE
Raise 400 Bad Request on server-side `await request.json()` if incorrect content-type received

### LINKED ISSUE #2174: Insecure way to handle json requests
## Long story short

[BaseRequest.json](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_request.py#L499) coroutine returns request body directly, whose `Content-Type` should have been checked. 

## Expected behaviour

`Content-Type` should be ensured to be `application/json` so that one can make sure it is sent by ajax, since there are always csrf attacks while ajax can not be forged cross site.

## Actual behaviour

`Content-Type` is not checked. 

## Steps to reproduce

I setup the server using following code. 
``` python
#!/usr/bin/env python3

from aiohttp import web
import asyncio


app = web.Application()


async def index(request):
    if request.method == 'GET':
        return web.Response(text='Hello Aiohttp!')
    elif request.method == 'POST':
        j = await request.json()
        return web.json_response(j)


def setup_routes(app):
    app.router.add_route('*', '/', index)


if __name__ == '__main__':
    setup_routes(app)
    web.run_app(app, host='127.0.0.1', port=8080)
```

And I request the server with following command. 

```shell
# both of the following will work
curl -v -d '{"abc":2}' -H "Content-Type: application/json" http://127.0.0.1:8080
curl -v -d '{"abc":2}' http://127.0.0.1:8080  # This should fail.
```

Actually [flask](http://flask.pocoo.org/) handles json requests only with `Content-Type: application/json`

## Environment
```shell
> uname -a
Darwin bogon 16.7.0 Darwin Kernel Version 16.7.0: Thu Jun 15 17:36:27 PDT 2017; root:xnu-3789.70.16~2/RELEASE_X86_64 x86_64

> python3 --version  # installed with homebrew
Python 3.6.2

> python3 -c 'import aiohttp; print(aiohttp.__version__)'
2.0.7

> brew --version 
Homebrew 1.3.0
Homebrew/homebrew-core (git revision 51fde5; last commit 2017-08-04)

```

### PR DESCRIPTION
**Pull request by atemate**
## What do these changes do?

make server-side request fail with `HTTPBadRequest` on `await request.json()` with wrong content-type for json.

This PR is an alternative implementation of https://github.com/aio-libs/aiohttp/pull/2177.

## Are there changes in behavior for the user?
The following code now raises HTTP 400:
```
async def handler(request):
    await request.json()  # raises HTTP 400
```

## Related issue number

closes https://github.com/aio-libs/aiohttp/issues/2174
## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Review by asvetlov — APPROVED** [review]
I'm curious why `aiohttp.svg` is changed?

**Review by asvetlov — CHANGES_REQUESTED** [review]
LGTM except for logo file

File: aiohttp.svg [hunk 1]
```diff
@@ -0,0 +1,2144 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (webknjaz) [inline aiohttp.svg]: This is not used/needed anywhere. Seems totally unrelated.

File: aiohttp/helpers.py [hunk 2]
```diff
@@ -106,6 +106,8 @@ def all_tasks(
 
 coroutines._DEBUG = old_debug  # type: ignore
 
+json_re = re.compile(r'^application/(?:[\w.+-]+?\+)?json')
```
Comment (webknjaz) [inline aiohttp/helpers.py:109]: > `+?`

This is probably not what you want.

File: aiohttp/helpers.py [hunk 3]
```diff
@@ -360,6 +362,13 @@ def content_disposition_header(disptype: str,
     return value
 
 
+def is_expected_content_type(response_content_type: str,
+                             expected_content_type: str) -> bool:
+    if expected_content_type == 'application/json':
```
Comment (webknjaz) [inline aiohttp/helpers.py:367]: I think this should go to a constant. Now it's hardcoded all over the place which is not maintainable.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: aiohttp/helpers.py (hunk 3 above)
Comment (asvetlov) [reply aiohttp/helpers.py:367]: I disagree.
The string is better readable than constant.
Also, we never change this line in the future.
What maintenance problem are you talking about?

File: aiohttp/helpers.py (hunk 2 above)
Comment (asvetlov) [reply aiohttp/helpers.py:109]: This regexp pattern is used by client code for years.
What is wrong with it?

File: aiohttp/helpers.py (hunk 3 above)
Comment (atemate) [reply aiohttp/helpers.py:367]: I think, the general Python code style welcomes usage of some well-known constants as literals and not const variables. For example, everyone knows that 404 is the HTTP code for Not Found, as well as `application/json` is the name for a mime type that will not be changed in future. So I would leave it as a string literal (same as the default parameter of `content_type` in `ClientResponse.json()`)

File: aiohttp/helpers.py (hunk 2 above)
Comment (atemate) [reply aiohttp/helpers.py:109]: yes, sorry for not pointing out to the fact that we simply moved `json_re` from this place: https://github.com/aio-libs/aiohttp/pull/3889/files#diff-2fa8ab44317df4fbb7864536a6794d7aL83

File: aiohttp.svg (hunk 1 above)
Comment (atemate) [reply aiohttp.svg]: I'm totally sorry for this extra file, this is a mistake of mine :/

Comment (atemate) [reply aiohttp.svg]: deleted it.

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
@webknjaz please review again

File: aiohttp/helpers.py (hunk 2 above)
Comment (webknjaz) [reply aiohttp/helpers.py:109]: @asvetlov `+` means one or more, `?` means `0` or `1`. I don't get this syntax. Does it essentially mean `*` (`0` or more times)?

Comment (webknjaz) [reply aiohttp/helpers.py:109]: Oh, that's lazy mode. Nevermind.

**Review by webknjaz — APPROVED** [review]

**Comment by asvetlov** [comment]
thanks!
